### PR TITLE
Fix for issue #151 https://github.com/RestComm/jain-sip/issues/151

### PR DESCRIPTION
**Pull request by gfigiel**
Please see the fix for issue 151 - modified in-dialog provisional responses improperly marked as restransmitted.
#151

### CODE REVIEW

**Comment by gfigiel** [comment]
Could you please let me know when jain-sip-ri library release containing this fix is available? It is needed to modify the SIP11 RA dependency to fix the issue.

File: src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java [hunk 1]
```diff
@@ -72,6 +65,12 @@
 import javax.sip.header.RouteHeader;
 import javax.sip.header.TimeStampHeader;
 import javax.sip.message.Request;
+import java.io.IOException;
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java]: why do we need this additional imports?

File: src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java [hunk 2]
```diff
@@ -1545,7 +1544,7 @@ public void processResponse(SIPResponse sipResponse, MessageChannel incomingChan
     int code = sipResponse.getStatusCode();
     boolean isRetransmission = !responsesReceived.add(Integer.valueOf(code));
     if (code > 100 && code < 200 && isRetransmission) {
-      if (lastResponse != null && !sipResponse.toString().equals(lastResponse.toString())) {
+      if (lastResponse != null && !sipResponse.equals(lastResponse)) {
```
Comment (jaimecasero) [inline src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java]: Equals will compare the statusLine and the headers, not the body. In original issue is reported that "not deliver events to JSLEE application when provisional response (100 -200) is repeated with new content or header". Is this patch taking into account those "header" modification details...? I just see this fixing the "content" modifications...

Do you have a unit test to prove this working?

File: src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java (hunk 1 above)
Comment (gfigiel) [inline src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java]: I do not know why they are needed. This change was just my IDE imports optimizing. You can see that the only change here is the order of imports - no removal or additions of imports.

File: src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java (hunk 2 above)
Comment (gfigiel) [inline src/gov/nist/javax/sip/stack/SIPClientTransactionImpl.java]: As you noticed - equals() (in addtion to toStrong().equals() ) also compars the headers. Thus when the message content (ex. SDP will change) then the Content-Length header shall also change. I thik this shall be enough for the majority of cases. The general assumption is that if any header of the SIP response changes - it means that this is not a retransmitted response as allowed by SIP spec.

**Comment by gfigiel** [comment]
Hello, It seems that it is not the only needed change. I can see that the retransmission flag is set to true by default in SIPResponse:
https://github.com/RestComm/jain-sip/blob/45cd46d03e1e3504750b573dbc1871554c7fdbd4/src/gov/nist/javax/sip/message/SIPResponse.java#L58

Why it is done like that? It seems that in some cases this value is taken into consideration causing my fix is not enough...
I will still work on this issue...

**Comment by jaimecasero** [comment]
Original intention of the author to make isRetranmission true by default is unknown to me. You know when using boolean flags, sometimes it makes more sense(reduce the number of potential modifications) to start with true or not...

Having said that, i think DialogFilter is the class that mostly uses the isRetransmission info/flag. The SIPClientTransactionImpl is responsible to set the actual value of the flag during Response processing(as you already noted in the change that laters this logic). This should be run before reaching DialogFilter, and creating the proper ResponseEvent (where the flag is copied from original response message). The assumption is that Transaction layer is run before Dialog layer ...

**Pull request merged**